File: src/algebra1_multi_step_equations_hard.py
"""
algebra1 - multi_step_equations (hard)
Generated: 2026-02-22T03:49:47.535855
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 5)
    
    if problem_type == 1:
        # Type 1: Equations with variables on both sides and nested parentheses
        # e.g., 2(3x - 4) - 5(x + 1) = 3(2x - 1) + 7
        a1 = nonzero(-4, 4)
        b1 = nonzero(-6, 6)
        a2 = nonzero(-4, 4)
        b2 = nonzero(-6, 6)
        a3 = nonzero(-4, 4)
        b3 = nonzero(-6, 6)
        c = nonzero(-8, 8)
        
        left_side = a1*(a2*x + b1) + a3*(b2*x + b3)
        right_side = c
        
        left_expanded = expand(left_side)
        solutions = solve(Eq(left_expanded, right_side), x)
        
        if solutions and len(solutions) == 1:
            ans = solutions[0]
            if ans.is_rational and abs(ans) < 100:
                left_latex = f"{a1}({a2}x {'+' if b1 >= 0 else ''}{b1}) {'+' if a3 >= 0 else ''}{a3}({b2}x {'+' if b3 >= 0 else ''}{b3})"
                
                return problem(
                    question=f"Solve for $x$: ${left_latex} = {c}$",
                    answer=ans,
                    difficulty=(1600, 1750),
                    topic="algebra1/multi_step_equations",
                    solution=steps(
                        f"Distribute: ${latex(left_expanded)} = {c}$",
                        f"Combine like terms and solve",
                        f"$x = {latex(ans)}$"
                    )
                )
    
    if problem_type == 2:
        # Type 2: Equations with fractions on both sides
        # e.g., (2x + 3)/4 - (x - 1)/3 = (x + 5)/6
        a1 = nonzero(-5, 5)
        b1 = nonzero(-8, 8)
        d1 = choice([2, 3, 4, 5, 6])
        
        a2 = nonzero(-5, 5)
        b2 = nonzero(-8, 8)
        d2 = choice([2, 3, 4, 5, 6])
        if d2 == d1:
            d2 = choice([d for d in [2, 3, 4, 5, 6] if d != d1])
        
        a3 = nonzero(-5, 5)
        b3 = nonzero(-8, 8)
        d3 = choice([2, 3, 4, 5, 6])
        
        equation = Eq((a1*x + b1)/d1 - (a2*x + b2)/d2, (a3*x + b3)/d3)
        solutions = solve(equation, x)
        
        if solutions and len(solutions) == 1:
            ans = solutions[0]
            if ans.is_rational and abs(ans) < 100:
                return problem(
                    question=f"Solve for $x$: $\\frac{{{a1}x {'+' if b1 >= 0 else ''}{b1}}}{{{d1}}} - \\frac{{{a2}x {'+' if b2 >= 0 else ''}{b2}}}{{{d3}}} = \\frac{{{a3}x {'+' if b3 >= 0 else ''}{b3}}}{{{d3}}}$",
                    answer=ans,
                    difficulty=(1650, 1800),
                    topic="algebra1/multi_step_equations",
                    solution=steps(
                        f"Multiply through by LCD to clear fractions",
                        f"Simplify and combine like terms",
                        f"$x = {latex(ans)}$"
                    )
                )
    
    if problem_type == 3:
        # Type 3: Complex distribution with collection of like terms
        # e.g., 3(2x - 1) - 4(x + 2) + 2(3x - 5) = 7x + 1
        a1 = nonzero(-5, 5)
        b1 = nonzero(-7, 7)
        c1 = nonzero(-5, 5)
        
        a2 = nonzero(-5, 5)
        b2 = nonzero(-7, 7)
        c2 = nonzero(-5, 5)
        
        a3 = nonzero(-5, 5)
        b3 = nonzero(-7, 7)
        c3 = nonzero(-5, 5)
        
        right_a = nonzero(-7, 7)
        right_b = nonzero(-10, 10)
        
        left_side = a1*(c1*x + b1) + a2*(c2*x + b2) + a3*(c3*x + b3)
        right_side = right_a*x + right_b
        
        left_expanded = expand(left_side)
        solutions = solve(Eq(left_expanded, right_side), x)
        
        if solutions and len(solutions) == 1:
            ans = solutions[0]
            if ans.is_rational and abs(ans) < 100:
                left_latex = f"{a1}({c1}x {'+' if b1 >= 0 else ''}{b1}) {'+' if a2 >= 0 else ''}{a2}({c2}x {'+' if b2 >= 0 else ''}{b2}) {'+' if a3 >= 0 else ''}{a3}({c3}x {'+' if b3 >= 0 else ''}{b3})"
                
                return problem(
                    question=f"Solve for $x$: ${left_latex} = {right_a}x {'+' if right_b >= 0 else ''}{right_b}$",
                    answer=ans,
                    difficulty=(1600, 1750),
                    topic="algebra1/multi_step_equations",
                    solution=steps(
                        f"Distribute all terms: ${latex(left_expanded)} = {right_a}x {'+' if right_b >= 0 else ''}{right_b}$",
                        f"Move all $x$ terms to one side",
                        f"$x = {latex(ans)}$"
                    )
                )
    
    if problem_type == 4:
        # Type 4: Nested parentheses with distribution
        # e.g., 2[3(x - 1) + 4] - 5[2(x + 3) - 1] = 8
        outer1 = nonzero(-4, 4)
        inner1_a = nonzero(-3, 3)
        inner1_b = nonzero(-6, 6)
        add1 = nonzero(-5, 5)
        
        outer2 = nonzero(-4, 4)
        inner2_a = nonzero(-3, 3)
        inner2_b = nonzero(-6, 6)
        add2 = nonzero(-5, 5)
        
        rhs = nonzero(-15, 15)
        
        left_side = outer1*(inner1_a*(x + inner1_b) + add1) + outer2*(inner2_a*(x + inner2_b) + add2)
        left_expanded = expand(left_side)
        
        solutions = solve(Eq(left_expanded, rhs), x)
        
        if solutions and len(solutions) == 1:
            ans = solutions[0]
            if ans.is_rational and abs(ans) < 100:
                return problem(
                    question=f"Solve for $x$: ${outer1}[{inner1_a}(x {'+' if inner1_b >= 0 else ''}{inner1_b}) {'+' if add1 >= 0 else ''}{add1}] {'+' if outer2 >= 0 else ''}{outer2}[{inner2_a}(x {'+' if inner2_b >= 0 else ''}{inner2_b}) {'+' if add2 >= 0 else ''}{add2}] = {rhs}$",
                    answer=ans,
                    difficulty=(1700, 1850),
                    topic="algebra1/multi_step_equations",
                    solution=steps(
                        f"Distribute inner parentheses first",
                        f"Then distribute outer brackets",
                        f"Simplify: ${latex(left_expanded)} = {rhs}$",
                        f"$x = {latex(ans)}$"
                    )
                )
    
    # Type 5: Equations with multiple fractions and parentheses
    # e.g., (3x - 2)/4 + (2x + 1)/3 = (x - 5)/2 + 1
    a1 = nonzero(-4, 4)
    b1 = nonzero(-8, 8)
    d1 = choice([2, 3, 4, 6])
    
    a2 = nonzero(-4, 4)
    b2 = nonzero(-8, 8)
    d2 = choice([2, 3, 4, 6])
    
    a3 = nonzero(-4, 4)
    b3 = nonzero(-8, 8)
    d3 = choice([2, 3, 4, 6])
    
    constant = nonzero(-5, 5)
    
    equation = Eq((a1*x + b1)/d1 + (a2*x + b2)/d2, (a3*x + b3)/d3 + constant)
    solutions = solve(equation, x)
    
    if solutions and len(solutions) == 1:
        ans = solutions[0]
        if ans.is_rational and abs(ans) < 100:
            return problem(
                question=f"Solve for $x$: $\\frac{{{a1}x {'+' if b1 >= 0 else ''}{b1}}}{{{d1}}} + \\frac{{{a2}x {'+' if b2 >= 0 else ''}{b2}}}{{{d2}}} = \\frac{{{a3}x {'+' if b3 >= 0 else ''}{b3}}}{{{d3}}} {'+' if constant >= 0 else ''}{constant}$",
                answer=ans,
                difficulty=(1700, 1900),
                topic="algebra1/multi_step_equations",
                solution=steps(
                    f"Find LCD and multiply through to clear fractions",
                    f"Distribute and combine like terms",
                    f"$x = {latex(ans)}$"
                )
            )
    
    # Fallback to simple harder problem
    a = nonzero(-5, 5)
    b = nonzero(-8, 8)
    c = nonzero(-5, 5)
    d = nonzero(-8, 8)
    e = nonzero(-10, 10)
    
    left = a*(b*x + c) + d*x
    left_expanded = expand(left)
    solutions = solve(Eq(left_expanded, e), x)
    ans = solutions[0]
    
    return problem(
        question=f"Solve for $x$: ${a}({b}x {'+' if c >= 0 else ''}{c}) {'+' if d >= 0 else ''}{d}x = {e}$",
        answer=ans,
        difficulty=(1600, 1700),
        topic="algebra1/multi_step_equations",
        solution=steps(
            f"Distribute: ${latex(left_expanded)} = {e}$",
            f"$x = {latex(ans)}$"
        )
    )

emit(generate())

File: src/algebra1_multi_step_equations_hard.jl
# algebra1 - multi_step_equations (hard)
# Generated: 2026-03-08T19:53:51.988770

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("algebra1/multi_step_equations")
    
    problem_type = rand(1:5)
    
    if problem_type == 1
        # Nested equations with fractions and distribution
        ans = randint(-15, 15)
        a = nonzero(-8, 8)
        b = nonzero(-6, 6)
        c = randint(-20, 20)
        d = nonzero(2, 5)
        e = randint(-15, 15)
        
        # Build: a(x - b)/d + c = e, where x = ans
        lhs = a*(x - b)//d + c
        rhs = e
        
        # Solve for what c and e should be
        lhs_at_ans = substitute(lhs, x => ans)
        rhs = lhs_at_ans
        
        problem(
            question="Solve for \\(x\\): \$$(tex(lhs)) = $(tex(rhs))\$",
            answer=ans,
            difficulty=(1850, 2000),
            solution=steps(
                sol("Given", lhs ~ rhs),
                "Subtract $(c) from both sides: \$$(tex(a*(x-b)//d)) = $(tex(rhs - c))\$",
                "Multiply both sides by $(d): \$$(tex(a*(x-b))) = $(tex(d*(rhs - c)))\$",
                "Divide by $(a): \$$(tex(x-b)) = $(tex(d*(rhs - c)//a))\$",
                "Add $(b): \$x = $(ans)\$",
                sol("Answer", ans)
            ),
            time=180
        )
        
    elseif problem_type == 2
        # Variable on both sides with complex distribution
        ans = randint(-20, 20)
        a = nonzero(-7, 7)
        b = randint(-12, 12)
        c = nonzero(-7, 7)
        d = randint(-12, 12)
        e = nonzero(-5, 5)
        f = randint(-10, 10)
        
        # Build: a(x - b) + c = d(x - e) + f
        lhs = expand(a*(x - b) + c)
        rhs = expand(d*(x - e) + f)
        
        # Solve backwards
        # a*x - a*b + c = d*x - d*e + f
        # (a-d)*x = -d*e + f + a*b - c
        target_rhs = substitute(rhs, x => ans)
        c_new = substitute(lhs, x => ans) - target_rhs
        lhs = expand(a*(x - b)) + c_new
        
        problem(
            question="Solve for \\(x\\): \$$(tex(lhs)) = $(tex(rhs))\$",
            answer=ans,
            difficulty=(1900, 2100),
            solution=steps(
                sol("Given", lhs ~ rhs),
                "Expand both sides",
                "Move all \\(x\\) terms to one side and constants to the other",
                "Combine like terms and isolate \\(x\\)",
                sol("Answer", ans)
            ),
            time=210
        )
        
    elseif problem_type == 3
        # Absolute value equation with two solutions
        inner_offset = randint(-8, 8)
        dist = randint(5, 18)
        r1 = inner_offset - dist
        r2 = inner_offset + dist
        a = nonzero(-6, 6)
        
        # |a*x - a*inner_offset| = a*dist (or adjusted)
        inner = a*x - a*inner_offset
        
        problem(
            question="Solve for \\(x\\): \$|$(tex(inner))| = $(abs(a)*dist)\$",
            answer=fmt_set(Set([r1, r2])),
            difficulty=(1800, 2000),
            answer_type="set",
            solution=steps(
                sol("Given", "|$(tex(inner))| = $(abs(a)*dist)"),
                "Split into two cases:",
                "Case 1: \$$(tex(inner)) = $(abs(a)*dist)\$ gives \\(x = $(r2)\\)",
                "Case 2: \$$(tex(inner)) = -$(abs(a)*dist)\$ gives \\(x = $(r1)\\)",
                "Solution set: \$$(fmt_set(Set([r1, r2])))\$"
            ),
            time=180
        )
        
    elseif problem_type == 4
        # Rational equation requiring LCD
        ans = nonzero(-15, 15)
        a = nonzero(-8, 8)
        b = randint(-10, 10)
        c = nonzero(2, 6)
        d = nonzero(2, 6)
        
        # a/(c) + x/d = b, solve for x = ans
        # x/d = b - a/c
        # x = d*(b - a/c) = ans
        target = ans
        b = a//c + target//d
        
        problem(
            question="Solve for \\(x\\): \$\\frac{$(a)}{$(c)} + \\frac{x}{$(d)} = $(tex(b))\$",
            answer=ans,
            difficulty=(1950, 2150),
            solution=steps(
                sol("Given", "\\frac{$(a)}{$(c)} + \\frac{x}{$(d)} = $(tex(b))"),
                "Subtract \\(\\frac{$(a)}{$(c)}\\) from both sides",
                "Simplify: \\(\\frac{x}{$(d)} = $(tex(b - a//c))\\)",
                "Multiply by $(d): \\(x = $(ans)\\)",
                sol("Answer", ans)
            ),
            time=180
        )
        
    else
        # Complex multi-step with nested parentheses
        ans = randint(-18, 18)
        a = nonzero(-6, 6)
        b = nonzero(2, 5)
        c = randint(-12, 12)
        d = nonzero(-8, 8)
        e = randint(-15, 15)
        
        # a[b(x - c) - d] = e, solve backwards
        # b(x - c) - d = e/a
        # b(x - c) = e/a + d
        # x - c = (e/a + d)/b = (e + a*d)/(a*b)
        # x = c + (e + a*d)/(a*b) = ans
        
        e = a*b*ans - a*b*c - a*d
        
        inner = b*(x - c) - d
        lhs = a*inner
        
        problem(
            question="Solve for \\(x\\): \$$(tex(lhs)) = $(e)\$",
            answer=ans,
            difficulty=(2000, 2200),
            solution=steps(
                sol("Given", lhs ~ e),
                "Divide both sides by $(a): \$$(tex(inner)) = $(tex(e//a))\$",
                "Add $(d): \$$(tex(b*(x-c))) = $(tex(e//a + d))\$",
                "Divide by $(b): \$$(tex(x-c)) = $(tex((e//a + d)//b))\$",
                "Add $(c) to get \\(x = $(ans)\\)",
                sol("Answer", ans)
            ),
            time=240
        )
    end
end